service/history/transferQueueActiveTaskExecutor.go

```diff
@@ -319,11 +323,15 @@ func (t *transferQueueActiveTaskExecutor) processCloseExecution(
 	namespaceName := mutableState.GetNamespaceEntry().Name()
 	children := copyChildWorkflowInfos(mutableState.GetPendingChildExecutionInfos())
 
-	// NOTE: do not access anything related mutable state after this lock release
-	// release the context lock since we no longer need mutable state builder and
-	// the rest of logic is making RPC call, which takes time.
-	release(nil)
-	err = t.recordWorkflowClosed(
+	// NOTE: do not access anything related mutable state after this lock release.
+	// Lock is needed if execution is about to be deleted. Otherwise, release lock immediately
+	// since mutable state builder is not needed and the rest of logic is RPC calls, which can take time.
+	if task.DeleteAfterClose {
+		defer func() { release(nil) }()
```

Comment:
This is not right. We cannot hold the lock while doing the archiving visibility and other operations. 
We will need to release the lock, process those operations, and then handle the deleteAfterClose as a separate task which would obtain the lock again.

Reply:
I changed it to the separate lock.